analyze-network-with-subnetwork-trace/src/main/java/com/esri/arcgismaps/sample/analyzenetworkwithsubnetworktrace/MainActivity.kt

```diff
@@ -0,0 +1,512 @@
+                    "For a working barrier condition, try \"Transformer Load\" Equal \"15\": " +
+                            it.message.toString()
+                )
+            }
+            // get the UtilityElementTraceResult
+            val elementTraceResult = utilityTraceResults.first() as UtilityElementTraceResult
+
+            showLoadingDialog(false)
+            AlertDialog.Builder(this@MainActivity).apply {
+                // set the result dialog title
+                setTitle("Trace result")
+                // show the element result count
+                setMessage(elementTraceResult.elements.count().toString() + " elements found.")
+            }.show()
+        }
+    }
+
+    /**
+     * Convert the given [expression] into a string.
+     */
+    private fun expressionToString(expression: UtilityTraceConditionalExpression): String? {
+        when (expression) {
+            // when the expression is a category comparison expression
+            is UtilityCategoryComparison -> {
+                return expression.category.name + " " + expression.comparisonOperator
+            }
+            // when the expression is an utility trace AND condition
+            is UtilityTraceAndCondition -> {
+                return expressionToString(expression.leftExpression) + " AND\n" + expressionToString(
+                    expression.rightExpression
+                )
+            }
+            // when the expression is an utility trace OR condition
+            is UtilityTraceOrCondition -> {
+                return expressionToString(expression.leftExpression) + " OR\n" + expressionToString(
+                    expression.rightExpression
+                )
+            }
+            // when the expression is an attribute comparison expression
+            is UtilityNetworkAttributeComparison -> {
+                // the name and comparison operator of the expression
+                val networkAttributeNameAndOperator = expression.networkAttribute.name + " " +
+                        expression.comparisonOperator::class.simpleName + " "
+
+                // check whether the network attribute has a coded value domain
+                (expression.networkAttribute.domain as? CodedValueDomain)?.let { codedValueDomain ->
+                    // if there's a coded value domain name
+                    val codedValueDomainName = codedValueDomain.codedValues.first { codedValue ->
+                        codedValue.code?.let { code ->
+                            convertToDataType(code,
+                                expression.networkAttribute.dataType)
+                        } ==
+                                expression.value?.let { value ->
+                                    convertToDataType(
+                                        value,
+                                        expression.networkAttribute.dataType
+                                    )
+                                }
```

Comment:
could you refactor the block code into a function
```
codedValue ->
                        codedValue.code?.let { code ->
                            convertToDataType(code,
                                expression.networkAttribute.dataType)
                        } ==
                                expression.value?.let { value ->
                                    convertToDataType(
                                        value,
                                        expression.networkAttribute.dataType
                                    )
                                }
```
otherwise, it is hard to understand